contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,317 @@
+planned temporary process failures (disruptions) that an application 
+can tolerate. 
+
+When we expect a terminated Pod to be rescheduled immediately, or nearly so, 
+redistributing its data will only increase the expected time to recovery. This 
+is only exacerbated by the fact that when the terminated Pod is resurrected, 
+the application will have to once again redistribute its data to 
+make use of the restored capacity. For applications that may require TiBs of 
+data to be redistributed for a single Pod update, the network, storage, and 
+time costs associated with redistribution are prohibitive when there is a 
+very high probability that we can reschedule the Pod immediately. 
+
+If the application's Persistent Volumes are backed by local storage media (note 
+that this is not currently possible but we expect this feature to land in the 
+near future), we expect a decommissioned instance to behave identically to an 
+instance whose PVs are backed by remote storage media. However, as a local PV is 
+intrinsically tied to the lifetime of the Node, leaving the application 
+under replicated during destructive updates to the Pod is inherently more 
+risky. We still believe that, under these circumstances, temporary 
+under replication is a viable strategy, and users can mitigate their risk by 
+increasing the number of replicas of the application's data. 
+
+## Affected Components
+1. API Server
+1. API Machinery
+1. Controllers (that wish to utilize this feature)
+1. Kubelet
+1. Kubectl
+
+## Use Cases
+
+1. As the implementor of a controller or operator that manages a stateful 
+workload, I want to send a signal to the Pods created by my controller or 
+operator to control the application's replication behavior during Pod 
+termination.
+1. As the implementor of tenant applications on a Kubernetes cluster, I want 
+my applications to receive a signal when they are evicted from a node due to 
+intolerable taints.
+1. As the administrator of a Kubernetes cluster, I want to provide a signal to 
+tenant applications that they are being terminated in violation of their 
+declared disruption budgets or termination grace periods due to a superseding, 
+global, organization policy (e.g. My organization requires that an application's 
+termination grace period and disruption budget allow managed upgrades to make 
+progress.)
+1. As a user I want to be able to send a termination reason to my Pods when I 
+use the Kubernetes API or kubectl to gracefully delete a Pod.
+1. As a Reliability Engineer for an application running on a Kubernetes cluster,
+I want to configure Pods to communicate the reason that they were terminated via 
+a HTTP GET request so that I can configure monitoring and alerting based on 
+historic trends.
+
+## Assumptions
+
+1. Kubelet will always invoke the pre-stop handler prior to sending a `TERM` 
+signal to the entry point of a container.
+1. pre-stop handlers will not contain complicated or long running business logic. 
```

Comment:
I don't think this is the case, even today.  I've seen insanely complex / business logic pre-stop handlers.

Reply:
@kow3ns  why is that an assumption ? Would the termination reason not work if this assumption is not met ?